lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Pat.hs

```diff
@@ -55,33 +64,106 @@ data Pat
   | AppP Pat Pat {- Pattern for Type applications -}
   | {- This last one is a bit special. This represents a complete type declaration.
        The first Pat should be instantiated to `Name l` where l is a concrete name.
-       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains Names.
+       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains TyVarPs.
        The final Pat should be instantiated to a Pat body.
        In some languages, parts of this may be ignored. E.g. in Rust the type name doesn't matter (we use the constr name of the
        outermost inner sum for constructing types). -}
     DecP Pat Pat Pat
   deriving stock (Show, Eq, Ord)
 
-infixr 5 :*
+-- infixr 5 :*
 
 {- Utility functions. Turn a list of types into a product/record/sum type.
 -}
-toProd :: [Pat] -> Pat
-toProd = ProdP . foldr (:*) Nil
+toProdP :: [Pat] -> Pat
+toProdP = ProdP . foldr ConsP NilP
 
-toRec :: [Pat] -> Pat
-toRec = RecP . foldr (:*) Nil
+toRecP :: [Pat] -> Pat
+toRecP = RecP . foldr ConsP NilP
 
-toSum :: [Pat] -> Pat
-toSum = SumP . foldr (:*) Nil
+toSumP :: [Pat] -> Pat
+toSumP = SumP . foldr ConsP NilP
 
 {- Converts a pattern that consists of a well formed pattern list
    (i.e. patterns formed from :* and Nil) into a list of patterns.
 -}
 patList :: Pat -> Maybe [Pat]
 patList = \case
-  Nil -> Just []
-  p1 :* p2 -> (p1 :) <$> patList p2
+  NilP -> Just []
+  p1 `ConsP` p2 -> (p1 :) <$> patList p2
+  _ -> Nothing
+
+data Exp
+  = LitE Literal
+  | NilE
+  | ConsE Exp Exp
+  | LabelE Exp Exp
+  | RecE Exp
+  | ProdE Exp
+  | SumE Exp
+  | -- NO EXPRESSION VARS! EXPRESSIONS DON'T HAVE HOLES!
+    RefE Exp Exp {- 1st arg should be a ModuleName  -}
+  | AppE Exp Exp {- Pattern for Type applications -}
+  | DecE Exp Exp Exp
+  deriving stock (Show, Eq, Ord)
```

Comment:
So, in theory yes, in practice probably not. 

`Pat` and `Exp` represent... `[TyName,ConstrName,ModuleName,VarName,FieldName,ClassName,TyVar,Ty,TyApp,ForeignRef,LocalRef,TyDef,TyAbs,TyBody,Constructor,Sum,Field,Record,Tuple,Product]`

And so every function that has `Pat` and `Exp` in the signature would have to be replaced by a function that uses `Pat` and `a` where `a` is constrained by some `IsLBExp` class. Or we'd have to create an `Exp`-like GADT that "flattens" all of those ProtoCompat types into a single recursive thing. 

Note that I don't even think we can fix this with `Control.Unification` because the `t` structure over which variables are parameterized has to be a single type to be used in the way we want it. 

This, by the way, is _exactly_ why I tried so hard to convince you to use a flat/recursive type in the proto months ago. Though even if you somehow think that's a good idea now, it'd be too much work to switch over, and I'm not 100% sure that it would even really be possible w/ the limitations of proto buffers.

Reply:
Realistically I don't think you'd be happy with what we'd have to do in order to make this work w/ the PC types. If you're dead-set on it then I'll cast my type magic.... but for now I'm marking this as resolved. If you have an idea of how to do it simply then feel free to show me a brainstorm of it.